src/Elf.cpp

```diff
@@ -0,0 +1,989 @@
+
+        if (s->get_type() != Symbol::STT_NOTYPE) {
+            return false;
+        }
+
+        return linker->needs_plt_entry(r);
+    };
+
+    // We need to build the PLT, so it can be positioned along with
+    // the rest of the text sections.
+    Section plt(".plt", Section::SHT_PROGBITS);
+    plt.set_alignment(16);
+    plt.set_flags(Section::SHF_ALLOC | Section::SHF_EXECINSTR);
+    std::list<Symbol> plt_symbols;
+    std::map<const Symbol *, const Symbol *> plt_defs;
+    // Hack: We're defining the global offset table, so it shouldn't be treated as an external symbol.
+    plt_defs[&got_sym] = &got_sym;
+    for (const Section &s : obj.sections()) {
+        for (const Relocation &r : s.relocations()) {
+            if (!needs_plt_entry(r)) {
+                continue;
+            }
+
+            const Symbol *sym = get_symbol(r);
+            const Symbol *& plt_def = plt_defs[sym];
+            if (plt_def) {
+                // We already made a PLT entry for this symbol.
+                continue;
+            }
+
+            debug(2) << "Defining PLT entry for " << sym->get_name() << "\n";
+            plt_symbols.push_back(linker->add_plt_entry(*sym, plt, got, got_sym));
+
+            plt_def = &plt_symbols.back();
+            symbols[plt_def] = plt_def;
+        }
+    }
+
+    // Start placing the sections into the shared object.
+
+    // Leave room for the header, and program headers at the beginning of the file.
+    append_zeros(output, sizeof(ehdr));
+    append_zeros(output, sizeof(phdrs[0])*3);
+
+    // We need to perform the relocations. To do that, we need to position the sections
+    // where they will go in the final shared object.
+    write_section(plt);
+    for (const Section &s : obj.sections()) {
+        if (s.is_alloc() && !s.is_writable()) {
+            write_section(s);
+        }
+    }
+    append_padding(output, 4096);
+    text_phdr.p_filesz = output.size() - text_phdr.p_offset;
+
+    data_phdr.p_type = PT_LOAD;
+    data_phdr.p_flags = PF_W | PF_R;
+    safe_assign(data_phdr.p_offset, output.size());
+    data_phdr.p_align = 4096;
+    for (const Section &s : obj.sections()) {
+        if (s.is_alloc() && s.is_writable()) {
+            write_section(s);
+        }
+    }
+
+    // The got will be written again later, after we add entries to it.
+    write_section(got);
+
+    /// Now that we've written the sections that define symbols, we
+    // can generate the symbol table.
+    Section symtab(".symtab", Section::SHT_SYMTAB);
+    symtab.set_alignment(4);
+    symtab.set_flag(Section::SHF_ALLOC);
+    std::vector<Sym<T>> syms;
+    Sym<T> undef_sym;
+    memset(&undef_sym, 0, sizeof(undef_sym));
+    syms.push_back(undef_sym);
+    // TODO: This map with pointers as the key will lead to non-deterministic builds...
+    std::map<const Symbol *, uint16_t> symbol_idxs;
+    uint64_t local_count = 0;
+    for (bool is_local : {true, false}) {
+        for (const auto &i : symbols) {
```

Comment:
I was adding these so the disassembly was more readable, I noticed that without the local symbols, the disassembly was just a giant blob. It's fuzzy to be now, but I think the behavior I was seeing is that the disassembly used the .symtab to annotate the assembly with the symbols, and not the .dynsym. I think the thing I should do is make a .symtab with all the symbols, and the a .dynsym with just the linker relevant symbols.